mvc/models/model-binding.rst

```diff
@@ -1,8 +1,113 @@
+   :language: c#
+   :lines: 63-76  
+   
+.. note:: The strings in the URL are not case sensitive.
+
+Routing knows which action method to use, so it checks for model binding attributes applied to the action method's arguments. If it finds a model binding attribute it will use the data source specified by the attribute. Otherwise, for each argument (a primitive type for now, more on complex types later), it looks in its data sources for something with a matching name. In the above example the only argument is an integer named `id`, so that is what the binder will look for. Below is a list of the data sources in the order that the model binder looks through them:
+ 
+#. ``Request.Form``: These are form values that go in the HTTP request using the POST method.
+#. ``RouteData.Values``: Contains information about each segment of the route. Route segments are separated by a forward slash ``/``.
+#. ``Request.QueryStrings``: Data to the right of a question mark (?) in a URL, separated by ampersands (&). For example, genre and subgenre are query strings: http://contoso.com/movies?genre=Drama&subgenre=Crime
+
+.. note:: Form, RouteData, and QueryStrings are all stored as name-value pairs.
+
+In the example, the model binder is loos through the Request.Form first for a key named `id`. In this case it didn't find anything, so it moved on to RouteData.Values looking for a key with a name of `id`. In our example, it's a match, since the last portion of the route has a key value of `id`, and it is an integer. Binding happens, and the value of the id parameter becomes 2. If the action method's parameter were a complex type such as the ``Movie`` type, which likely contains both primitive and complex types, MVC's model binder will still handle it nicely. It uses reflection and recursion to traverse the properties of complex types through its tree, and comparing any primitive members' names against the names of keys in data sources. The model binder can bind objects that contain nested objects for multiple levels. 
+
+.. note:: When the binder finds the first matching name in the data source, it stops looking for more elements matching that name, and moves onto the next property or item in the parameter list. 
+ 
+.. note:: If for some reason binding can't happen, it quietly fails.
+ 
+Additionally, there are some special data types that ASP.NET MVC must consider when performing model binding:
+
+- ``Request.Files``: Uploaded files that are part of the HTTP request.
+- ``IFormCollection``: The collection of key-value pairs that represent the form data.
+- Cancelation token: Used to cancel activity in asynchronous controllers.
+
+Once model binding is complete, validation occurs. The default model binders are quite smart, and work great for the vast majority of development scenarios. ASP.NET MVC is extensible so if you have unique needs you can customize the built in model binders.  
+
+Customize default model binding behavior with attributes 
```

Comment:
I think that before showing how to customize model binding it would be valuable to show at least one "complex" example (e.g. `Movie`) to drill in that model binding can do complex types. Nothing fancy - just movie id/title/genre.